pkg/cli/zip.go

```diff
@@ -0,0 +1,268 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+
+package cli
+
+import (
+	"archive/zip"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"sort"
+	"strings"
+
+	"github.com/cockroachdb/cockroach/pkg/server/serverpb"
+	"github.com/spf13/cobra"
+)
+
+var debugZipCmd = &cobra.Command{
+	Use:   "zip [file]",
+	Short: "gather cluster debug data into a zip file",
+	Long: `
+
+Gather cluster debug data into a zip file. Data includes cluster events, node
+liveness, node status, range status, node stack traces, log files, and SQL
+schema.
+
+Retrieval of per-node details (status, stack traces, range status) requires the
+node to be live and operating properly. Retrieval of SQL data requires the
+cluster to be live.
+`,
+	RunE: MaybeDecorateGRPCError(runDebugZip),
+}
+
+type zipper struct {
+	f *os.File
+	z *zip.Writer
+}
+
+func newZipper(f *os.File) *zipper {
+	return &zipper{
+		f: f,
+		z: zip.NewWriter(f),
+	}
+}
+
+func (z *zipper) close() {
+	z.z.Close()
+	z.f.Close()
+}
+
+func (z *zipper) create(name string) (io.Writer, error) {
+	fmt.Printf("  %s\n", name)
+	return z.z.Create(name)
+}
+
+func (z *zipper) createRaw(name string, b []byte) error {
+	w, err := z.create(name)
+	if err != nil {
+		return err
+	}
+	_, err = w.Write(b)
+	return err
+}
+
+func (z *zipper) createJSON(name string, m interface{}) error {
+	b, err := json.MarshalIndent(m, "", "  ")
+	if err != nil {
+		return err
+	}
+	return z.createRaw(name, b)
+}
+
+func (z *zipper) createError(name string, e error) error {
+	fmt.Printf("  %s: %s\n", name, e)
+	w, err := z.z.Create(name)
+	if err != nil {
+		return err
+	}
+	fmt.Fprintf(w, "%s\n", e.Error())
+	return nil
+}
+
+func runDebugZip(cmd *cobra.Command, args []string) error {
+	const (
+		base         = "debug"
+		eventsName   = base + "/events"
+		livenessName = base + "/liveness"
+		nodesPrefix  = base + "/nodes"
+		schemaPrefix = base + "/schema"
+	)
+
+	if len(args) != 1 {
+		return errors.New("output file argument is required")
+	}
+
+	conn, stopper, err := getGRPCConn()
+	if err != nil {
+		return err
+	}
+	defer stopper.Stop()
+
+	status := serverpb.NewStatusClient(conn)
+	admin := serverpb.NewAdminClient(conn)
+	ctx := stopperContext(stopper)
+
+	name := args[0]
+	if !strings.HasSuffix(name, ".zip") {
+		name += ".zip"
+	}
+	out, err := os.Create(name)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("writing %s\n", name)
+
+	z := newZipper(out)
+	defer z.close()
+
+	if events, err := admin.Events(ctx, &serverpb.EventsRequest{}); err != nil {
```

Comment:
maybe add some logging about attempt/success/failure of each item?
eg:
```
events: OK
liveness: FAIL (error: blah blah)
```

this would be clearer than failures being logged silently to the corresponding file. We could include that summary in the zip too.